### PR TITLE
Use System.Buffers + ValueTask in FormReading

### PR DESCRIPTION
**Pull request by benaadams**
Built on #550
Addressing Task allocations in  #553 (last commit in this PR)

ValueTask reduces Task allocation by FormReader to 1%
Post for 30,000 requests - byte[] allocations are as follows 0.5MB on rent vs 440MB on MemoryStream originally

/cc @rynowak @davidfowl @Tratcher @stephentoub

### CODE REVIEW

**Comment by stephentoub** [comment]
> This is because  ValueTask  isn't supported in net451

Why? The main point of putting it in a new assembly was so it would be usable there:
https://github.com/dotnet/corefx/blob/master/src/System.Threading.Tasks.Extensions/src/System.Threading.Tasks.Extensions.csproj#L9

**Comment by benaadams** [comment]
> > This is because ValueTask isn't supported in net451

> Why? The main point of putting it in a new assembly was so it would be usable there:

Entirely my fault as I couldn't follow instructions; when it said it needed `"System.Threading.Tasks": "4.0.0.0"` it literally meant exactly that, not  `"System.Threading.Tasks": "4.0.0-*"`

**Comment by benaadams** [comment]
Current

current

Buffers + ValueTask

valuetask

**Comment by benaadams** [comment]
Reduces Task allocation by FormReader to 1% https://github.com/aspnet/HttpAbstractions/issues/553#issuecomment-182826023

**Comment by benaadams** [comment]
Pre-for 3,000 requests 44MB on MemoryStream

pre-alloc

Post for 30,000 requests - `byte[]` allocations are as follows 0.5MB on rent vs 440MB on MemoryStream originally

allocs

**Comment by rynowak** [comment]
Nice :+1:

**Comment by benaadams** [comment]
Rebased; removed WebEncoders changes as its changed somewhat.

**Comment by SajayAntony** [comment]
Sweet :+1: /cc @DavidObando @mnltejaswini

**Comment by benaadams** [comment]
Need to sort out some `System.Runtime` references :-/

**Comment by DavidObando** [comment]
Love this.

File: src/Microsoft.AspNetCore.WebUtilities/FileBufferingReadStream.cs [hunk 1]
```diff
@@ -43,6 +56,17 @@ public class FileBufferingReadStream : Stream
                 throw new ArgumentNullException(nameof(tempFileDirectoryAccessor));
             }
 
+            _bytePool = bytePool;
+            if (memoryThreshold < 1024 * 1024)
```
Comment (rynowak) [inline src/Microsoft.AspNetCore.WebUtilities/FileBufferingReadStream.cs]: Would be nice to put this in a constant.

File: src/Microsoft.AspNetCore.WebUtilities/FileBufferingReadStream.cs [hunk 2]
```diff
@@ -17,11 +18,14 @@ namespace Microsoft.AspNetCore.WebUtilities
...
         private readonly Stream _inner;
+        private readonly ArrayPool<byte> _bytePool;
         private readonly int _memoryThreshold;
         private string _tempFileDirectory;
         private readonly Func<string> _tempFileDirectoryAccessor;
 
-        private Stream _buffer = new MemoryStream(); // TODO: We could have a more efficiently expanding buffer stream.
+        private Stream _buffer;
+        private byte[] _rentedBuffer;
+        private int _readIntoBuffer = 0;
```
Comment (rynowak) [inline src/Microsoft.AspNetCore.WebUtilities/FileBufferingReadStream.cs]: Default init not needed

File: src/Microsoft.AspNetCore.WebUtilities/FormReader.cs [hunk 3]
```diff
@@ -44,6 +61,9 @@ public FormReader(Stream stream, Encoding encoding)
                 throw new ArgumentNullException(nameof(encoding));
             }
 
+            _buffer = charPool.Rent(8192);
```
Comment (rynowak) [inline src/Microsoft.AspNetCore.WebUtilities/FormReader.cs]: Would be nice to put this in a constant - same with the buffer for streamreader

File: src/Microsoft.AspNetCore.WebUtilities/FormReader.cs [hunk 4]
```diff
@@ -186,7 +263,7 @@ private void Buffer()
         /// <param name="stream">The HTTP form body to parse.</param>
         /// <param name="cancellationToken">The <see cref="CancellationToken"/>.</param>
         /// <returns>The collection containing the parsed HTTP form body.</returns>
-        public static Task<Dictionary<string, StringValues>> ReadFormAsync(Stream stream, CancellationToken cancellationToken = new CancellationToken())
+        public static ValueTask<Dictionary<string, StringValues>> ReadFormAsync(Stream stream, CancellationToken cancellationToken = default(CancellationToken))
```
Comment (rynowak) [inline src/Microsoft.AspNetCore.WebUtilities/FormReader.cs]: What's the impact of putting `ValueTask<>` in a public API - can mere mortals still call this method?

File: src/Microsoft.AspNetCore.WebUtilities/FormReader.cs [hunk 5]
```diff
@@ -198,19 +275,69 @@ private void Buffer()
...
             }
+        }
 
-            return accumulator.GetResults();
+        private static async Task<Dictionary<string, StringValues>> ReadFormAsyncAwaited(Task<KeyValuePair<string, string>?> pairTask, FormReader reader, KeyValueAccumulator accumulator, CancellationToken cancellationToken)
+        {
+            using (reader)
+            {
+                var pair = await pairTask;
+                while (pair.HasValue)
```
Comment (rynowak) [inline src/Microsoft.AspNetCore.WebUtilities/FormReader.cs]: how does this work? Am I missing something that way over my head?

File: src/Microsoft.AspNetCore.WebUtilities/HttpRequestStreamReader.cs [hunk 6]
```diff
@@ -34,40 +34,13 @@ public class HttpRequestStreamReader : TextReader
         private bool _isBlocked;
 
         public HttpRequestStreamReader(Stream stream, Encoding encoding)
-            : this(stream, encoding, DefaultBufferSize)
+            : this(stream, encoding, DefaultBufferSize, ArrayPool<byte>.Shared, ArrayPool<char>.Shared)
```
Comment (rynowak) [inline src/Microsoft.AspNetCore.WebUtilities/HttpRequestStreamReader.cs]: :heart_eyes_cat:

File: src/Microsoft.AspNetCore.WebUtilities/StreamHelperExtensions.cs [hunk 7]
```diff
@@ -9,14 +10,25 @@ namespace Microsoft.AspNetCore.WebUtilities
 {
     public static class StreamHelperExtensions
```
Comment (rynowak) [inline src/Microsoft.AspNetCore.WebUtilities/StreamHelperExtensions.cs]: Do we really have these as public extensions on stream?

File: src/Microsoft.AspNetCore.WebUtilities/FormReader.cs [hunk 8]
```diff
@@ -44,6 +61,9 @@ public FormReader(Stream stream, Encoding encoding)
                 throw new ArgumentNullException(nameof(encoding));
             }
 
+            _buffer = charPool.Rent(8192);
+            _charPool = charPool;
+            _bytePool = bytePool;
             _reader = new StreamReader(stream, encoding, detectEncodingFromByteOrderMarks: true, bufferSize: 1024 * 2, leaveOpen: true);
```
Comment (rynowak) [inline src/Microsoft.AspNetCore.WebUtilities/FormReader.cs]: Should we use `HttpRequestBodyStreamReader` here?

File: src/Microsoft.AspNetCore.WebUtilities/FormReader.cs (hunk 4 above)
Comment (benaadams) [inline src/Microsoft.AspNetCore.WebUtilities/FormReader.cs]: Lol, yes is fine and happily awaitable

File: src/Microsoft.AspNetCore.WebUtilities/StreamHelperExtensions.cs (hunk 7 above)
Comment (benaadams) [inline src/Microsoft.AspNetCore.WebUtilities/StreamHelperExtensions.cs]: Yes :)

File: src/Microsoft.AspNetCore.WebUtilities/FormReader.cs (hunk 8 above)
Comment (benaadams) [inline src/Microsoft.AspNetCore.WebUtilities/FormReader.cs]: ¯\\_(ツ)_/¯

**Comment by benaadams** [comment]
Rebased, somewhat squashed, feedback made

@rynowak you asked a question on one of the algos, but it seemed to get lost in the rebase; can you ask again?

**Comment by benaadams** [comment]
@rynowak found it in email; will reply in-line

File: src/Microsoft.AspNetCore.WebUtilities/FormReader.cs [hunk 9]
```diff
@@ -198,19 +276,69 @@ private void Buffer()
...
             }
+        }
 
-            return accumulator.GetResults();
+        private static async Task<Dictionary<string, StringValues>> ReadFormAsyncAwaited(Task<KeyValuePair<string, string>?> pairTask, FormReader reader, KeyValueAccumulator accumulator, CancellationToken cancellationToken)
+        {
+            using (reader)
+            {
+                var pair = await pairTask;
+                while (pair.HasValue)
```
Comment (benaadams) [inline src/Microsoft.AspNetCore.WebUtilities/FormReader.cs]: > how does this work? Am I missing something that way over my head?

@rynowak do you mean the splitting of the `async` version from the non-`async` version (until it becomes async) or the end of read being indicated by a `null`?

Comment (rynowak) [inline src/Microsoft.AspNetCore.WebUtilities/FormReader.cs]: I just hadn't read these two lines super well.

**Comment by davidfowl** [comment]
@benaadams Can you rebase?

File: src/Microsoft.AspNetCore.WebUtilities/FormReader.cs (hunk 9 above)
Comment (benaadams) [inline src/Microsoft.AspNetCore.WebUtilities/FormReader.cs]: I imagine a golden time in the future when the compiler automatically does these patterns for you...

**Comment by benaadams** [comment]
Rebased; note: I had to bump ` src/Microsoft.Net.Http.Headers/project.json` from `"netstandard1.0"` to `"netstandard1.1"`

File: src/Microsoft.AspNetCore.WebUtilities/project.json [hunk 10]
```diff
@@ -16,12 +16,14 @@
...
-    "System.Text.Encodings.Web": "4.0.0-*"
+    "System.Text.Encodings.Web": "4.0.0-*",
+    "System.Threading.Tasks.Extensions": "4.0.0-*"
   },
   "frameworks": {
     "net451": {
       "frameworkAssemblies": {
-        "System.Runtime": { "type": "build" }
+        "System.Runtime": { "type": "build" },
+        "System.Threading.Tasks": "4.0.0.0"
```
Comment (davidfowl) [inline src/Microsoft.AspNetCore.WebUtilities/project.json]: This needs to be type build like the other facades

Comment (benaadams) [inline src/Microsoft.AspNetCore.WebUtilities/project.json]: Done

File: src/Microsoft.AspNetCore.Owin/WebSockets/OwinWebSocketAdapter.cs [hunk 11]
```diff
@@ -126,11 +128,18 @@ public override Task SendAsync(ArraySegment<byte> buffer, WebSocketMessageType m
...
+            try
             {
-                // Drain until close received
-                await ReceiveAsync(new ArraySegment<byte>(buffer), cancellationToken);
+                while (State == WebSocketState.CloseSent)
+                {
+                    // Drain until close received
+                    await ReceiveAsync(new ArraySegment<byte>(buffer), cancellationToken);
+                }
+            }
```
Comment (muratg) [inline src/Microsoft.AspNetCore.Owin/WebSockets/OwinWebSocketAdapter.cs]: Would this ever actually throw?

Comment (benaadams) [inline src/Microsoft.AspNetCore.Owin/WebSockets/OwinWebSocketAdapter.cs]: Potentially, could throw if socket already disconnected or cancellation token cancelled.

File: src/Microsoft.AspNetCore.WebUtilities/FileBufferingReadStream.cs [hunk 12]
```diff
@@ -135,22 +180,27 @@ private Stream CreateTempFile()
         public override int Read(byte[] buffer, int offset, int count)
         {
             ThrowIfDisposed();
-            if (_buffer.Position < _buffer.Length || _completelyBuffered)
+            if (_buffer.Position < _readIntoBuffer || _completelyBuffered)
```
Comment (Tratcher) [inline src/Microsoft.AspNetCore.WebUtilities/FileBufferingReadStream.cs]: Why was this changed?

Comment (benaadams) [inline src/Microsoft.AspNetCore.WebUtilities/FileBufferingReadStream.cs]: The buffer pool can return arrays larger than you requested, so we use the requested length rather than the array length

Comment (Tratcher) [inline src/Microsoft.AspNetCore.WebUtilities/FileBufferingReadStream.cs]: MemoryStream.Length is not the length of the buffer, it's the length of the data. Contrast with Capacity which is the length of the buffer.

Comment (benaadams) [inline src/Microsoft.AspNetCore.WebUtilities/FileBufferingReadStream.cs]: Bleh, sorry has been a while since this PR :)

Using a fixed sized memory stream rather than a resizable one; length in the fixed size case is the size of the underlying array not its used space (different behavior between the two inits :-/)

Comment (Tratcher) [inline src/Microsoft.AspNetCore.WebUtilities/FileBufferingReadStream.cs]: Ah, it assumes the `byte[]` started full of data. You could call `SetLength(0)` just after creating the stream to reset it.

Comment (benaadams) [inline src/Microsoft.AspNetCore.WebUtilities/FileBufferingReadStream.cs]: Hmm... poked around in MemoryStream, looks like that would work; had some issues when trying to get it to work before but maybe I was changing capacity

**Comment by Tratcher** [comment]
This ValueTask pattern causes substantial code duplication and maintenance issues. I don't think it's piratical. Can we split this and just take the buffer pooling?

**Comment by benaadams** [comment]
The ValueTask does cut the allocations on 30k requests from 750MB to 7MB though

/cc @rynowak

**Comment by Tratcher** [comment]
The buffer pooling is ready to merge and unrelated to the ValueTask work. Let's get that merged.

We need to come up with a simpler pattern if we're going to use ValueTasks.

**Comment by benaadams** [comment]
K, will split them later today

**Comment by benaadams** [comment]
> We need to come up with a simpler pattern if we're going to use ValueTasks.

Know what you mean, there are a bunch of higher performance patterns with async that do make the code more confusing as you need to split the code into multiple fragments.

I wonder if there is scope to push this into the compiler instead? As it already does that for general async.

Would rosyln be the correct repo to raise this? /cc @stephentoub

**Comment by stephentoub** [comment]
> Would roslyn be the correct repo to raise this?

https://github.com/dotnet/roslyn/issues/7169

**Comment by Tratcher** [comment]
I recommend we close this pending the roslyn discussion. @benaadams @rynowak?

**Comment by rynowak** [comment]
That seems fine, I don't think we're going to take this PR in it's current form
